node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/Rebalancer.java

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.provision.maintenance;
+
+import com.yahoo.config.provision.NodeResources;
+import com.yahoo.config.provision.NodeType;
+import com.yahoo.transaction.Mutex;
+import com.yahoo.vespa.hosted.provision.Node;
+import com.yahoo.vespa.hosted.provision.NodeList;
+import com.yahoo.vespa.hosted.provision.NodeRepository;
+import com.yahoo.vespa.hosted.provision.node.Agent;
+import com.yahoo.vespa.hosted.provision.provisioning.DockerHostCapacity;
+import com.yahoo.vespa.hosted.provision.provisioning.HostResourcesCalculator;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+
+public class Rebalancer extends Maintainer {
+
+    private final HostResourcesCalculator hostResourcesCalculator;
+    private final Clock clock;
+
+    public Rebalancer(NodeRepository nodeRepository, HostResourcesCalculator hostResourcesCalculator, Clock clock, Duration interval) {
+        super(nodeRepository, interval);
+        this.hostResourcesCalculator = hostResourcesCalculator;
+        this.clock = clock;
+    }
+
+    @Override
+    protected void maintain() {
+        // Work with an unlocked snapshot as this can take a long time and full consistency is not needed
+        NodeList allNodes = nodeRepository().list();
+
+        if ( ! zoneIsStable(allNodes)) return;
+
+        Move bestMove = findBestMove(allNodes);
+        if (bestMove == Move.none) return;
+        markWantToRetire(bestMove.node);
+    }
+
+    private boolean zoneIsStable(NodeList allNodes) {
+        List<Node> active = allNodes.state(Node.State.active).asList();
+        if (active.stream().anyMatch(node -> node.allocation().get().membership().retired())) return false;
+        if (active.stream().anyMatch(node -> node.status().wantToRetire())) return false;
+        return true;
+    }
+
+    /**
+     * Find the best move to reduce allocation skew and returns it.
+     * Returns Move.none if no moves can be made to reduce skew.
+     */
+    private Move findBestMove(NodeList allNodes) {
+        DockerHostCapacity capacity = new DockerHostCapacity(allNodes, hostResourcesCalculator);
+        Move bestMove = Move.none;
+        for (Node node : allNodes.state(Node.State.active).asList()) {
```

Comment:
> Limit this to tenant nodes, then you can also drop the second if

I'd rather keep it there as long as the code otherwise supports tenant nodes without parents.

The other 2: Thanks, done.

Reply:
> I'd rather keep it there as long as the code otherwise supports tenant nodes without parents.

The first `if` allows for that already?

Anyway, limiting this to `tenant`s is a good idea anyway to prevent the `Rebalancer` from attempting to retiring a `proxy`/`configserver` node (they are technically overallocated in AWS, there should be skew).

Reply:
> limiting this to tenants is a good idea

Good point. Another PR incoming for that.

(The second if is to not consider moving a node to the same host.)

Reply:
https://github.com/vespa-engine/vespa/pull/11011